docs/tsp/typeServerProtocol.ts

```diff
@@ -0,0 +1,383 @@
+        TypeParam = 3, // A type parameter is a placeholder for a type that can be specified when the type is used, such as in generics.
+        TypeAlias = 4, // A type alias is a name that refers to another type, such as `List[int]` or `Dict[str, int]`.
+        Function = 5, // A function is a named block of code that can be called with arguments and returns a value.
+        Class = 6, // A class is a blueprint for creating objects that encapsulates data and behavior.
+        SpecialBuiltInClass = 7, // Special built-in types like 'Tuple', 'Generic',  'Protocol', 'Callable', 'Type', etc
+        Alias = 8, // An alias declaration, which is a reference to another declaration.
+    }
+
+    // Represents a symbol declaration in the type system.
+    // A declaration is a specific instance of a symbol in the source code, such as a variable, function, or class.
+    export interface Declaration {
+        // Unique identifier for the declaration within the session.
+        handle: string | number;
+
+        // Category of this symbol (function, variable, etc.).
+        category: DeclarationCategory;
+
+        // Parse node associated with the declaration
+        node?: Node;
+
+        // The dot-separated import name for the file that
+        // contains the declaration (may not be definitive
+        // because a source file can be accessed via different
+        // import names in some cases).
+        moduleName: string;
+    }
+
+    // Synthesized type information for a declaration that is not directly
+    // represented in the source code, but is derived from the declaration. Other languages
+    // may refer to this as an anonymous type.
+    export interface SynthesizedTypeInfo {
+        type: Type;
+
+        // An optional node that is not used by the type evaluator
+        // but can be used by language services to provide additional
+        // functionality (such as go-to-definition).
+        node?: Node;
+    }
+
+    // Declaration information for a symbol, which includes
+    // a list of declarations and potentially synthesized types for those declarations.
+    export interface SymbolDeclInfo {
+        decls: Declaration[];
+        synthesizedTypes: SynthesizedTypeInfo[];
+    }
+
+    // Describes an import
+    export interface ImportedModuleDescriptor {
+        leadingDots: number;
+        nameParts: string[];
+    }
+
+    // Options for resolving an alias declaration.
+    export interface ResolveAliasOptions {
+        resolveLocalNames?: boolean; // Whether to resolve local names in the alias declaration.
+        allowExternallyHiddenAccess?: boolean; // Whether to allow access to members that are hidden by external modules.
+        skipFileNeededCheck?: boolean; // Whether to skip checking if the file is needed for the alias resolution.
+    }
+
+    // Describes a file that's been opened in the language server which the type server needs to be aware of.
+    export interface TextDocumentOpenParams {
```

Comment:
What is the file ownership model proposed by the TSP? Does the type server own the contents of closed files? Who is responsible for detecting on-disk changes? Is that the type server or the language server?

Reply:
Hopefully everything will work the same as it does in LSP. The client provides the content of open files. The server reads closed files from disk. Changes to disk (to force the type server to update any internal caches) will come in the form of a notification.